### PR TITLE
UCP/TEST: Select RMA/AMO transport according to remote key.

### LINKED ISSUE #397: Malloc + UCP Mmap does not work
The issue was reported by OpenSHMEM developers.

### PR DESCRIPTION
**Pull request by yosefe**
Support multiple RMA and AMO transports per endpoint, and select the best available transport according to remote key.
Fixes #397 - For example, if xpmem is not present, RMA/AMO to shared memory allocated by sysv/mmap(posix) will revert to IB.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1236/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1846/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1238/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1852/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1853/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1239/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1241/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1855/ for details.

**Comment by shamisp** [comment]
@MattBBaker - can you try it on cray? It supposed to resolve the issue that you have been complaining about.

File: src/ucp/core/ucp_ep.inl [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+{
+    return (lane_map >> (lane * UCP_PD_INDEX_BITS)) & UCS_MASK(UCP_PD_INDEX_BITS);
+}
+
+static inline const char* ucp_ep_peer_name(ucp_ep_h ep)
+{
+#if ENABLE_DEBUG_DATA
+    return ep->peer_name;
+#else
+    return "??";
```
Comment (shamisp) [inline src/ucp/core/ucp_ep.inl]: @yosefe - maybe return something more informative ? No debug enabled ? Or enable debug for more info ?

File: src/ucp/core/ucp_ep.inl [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+#endif
+}
+
+static inline ucp_pd_lane_map_t ucp_ep_pd_map_expand(ucp_pd_map_t pd_map)
+{
+    /* TODO use SSE and support larger pd_lane_map with shuffle / broadcast */
+    UCS_STATIC_ASSERT(sizeof(ucp_pd_lane_map_t) == sizeof(uint64_t));
+    return pd_map * 0x0101010101010101ul;
```
Comment (shamisp) [inline src/ucp/core/ucp_ep.inl]: Can you please document this magic code :)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1857/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1243/ for details (Mellanox internal link).

File: src/ucp/core/ucp_context.h [hunk 3]
```diff
@@ -14,15 +14,33 @@
...
+typedef uint8_t                      ucp_rsc_index_t;
+
+/* PDs */
+#define UCP_UINT_TYPE(_bits)         typedef UCS_PP_TOKENPASTE(UCS_PP_TOKENPASTE(uint, _bits), _t)
+#define UCP_PD_INDEX_BITS            8  /* How many bits are in PD index */
+#define UCP_MAX_PDS                  (1ul << UCP_PD_INDEX_BITS)
+UCP_UINT_TYPE(UCP_PD_INDEX_BITS)     ucp_pd_map_t;
+
+/* Lanes */
+#define UCP_MAX_LANES                8
```
Comment (shamisp) [inline src/ucp/core/ucp_context.h]: does it mean that UCP_PD_INDEX_BITS must be equal to UCP_MAX_LANES ?
If so we have to link this two.

Comment (yosefe) [inline src/ucp/core/ucp_context.h]: no, but their product must be 64 (which is both documented and checked at compile time)

**Comment by MattBBaker** [comment]
This is generating all sorts of interesting problems on Cray. I'm trying to isolate them and pin down what all is going on and debugging it.

**Comment by MattBBaker** [comment]
So, all of the problem tests have a common routine, uct_mm_mkey_pack(). This seems to segfault pretty reliably.

```
[----------] 2 tests from shm/test_ucp_wireup_rma
[ RUN      ] shm/test_ucp_wireup_rma.one_sided_wireup/0
[nid03791:12336:0] Caught signal 11 (Segmentation fault: address not mapped to object at address 0xdeadbee0)
#0  0x00002aaaad4843bf in waitpid () from /lib64/libpthread.so.0
#1  0x00002aaaaad93eb1 in ucs_debugger_attach () at debug/debug.c:617
#2  0x00002aaaaad945cd in ucs_handle_error () at debug/debug.c:819
---Type <return> to continue, or q <return> to quit---
#3  0x00002aaaaad948ec in ucs_error_signal_handler (signo=11,
    info=0x7fffffff6430, context=<optimized out>) at debug/debug.c:808
#4  <signal handler called>
#5  0x00002aaaad9c5032 in __strcpy_ssse3 () from /lib64/libc.so.6
#6  0x00002aaaab21a0ae in uct_mm_mkey_pack (pd=<optimized out>,
    memh=<optimized out>, rkey_buffer=0x8c79ea) at sm/mm/mm_pd.c:140
#7  0x00002aaaab20cf7a in uct_pd_mkey_pack (pd=<optimized out>,
    memh=<optimized out>, rkey_buffer=rkey_buffer@entry=0x8c79e2)
    at base/uct_pd.c:425
#8  0x00002aaaab641698 in ucp_rkey_pack (context=0x8c3c00, memh=0x8c65c0,
    rkey_buffer_p=0x7fffffff69c8, size_p=0x7fffffff69d0) at core/ucp_rkey.c:68
#9  0x000000000050d2ad in test_ucp_wireup_rma::rma_send (
    this=this@entry=0x8d3f80, from=0x8c54c0, to=to@entry=0x927040)
    at ucp/test_ucp_wireup.cc:259
#10 0x000000000050d900 in test_ucp_wireup_rma_one_sided_wireup_Test::test_body
    (this=0x8d3f80) at ucp/test_ucp_wireup.cc:285
#11 0x00000000004368b6 in ucs::test_base::run (this=0x8d3f80)
    at common/test.cc:140
#12 0x0000000000436c2d in ucs::test_base::TestBodyProxy (this=0x8c7a02)
    at common/test.cc:166
#13 0x00000000004bad85 in ucp_test::TestBody (this=<optimized out>)
    at ucp/ucp_test.h:27
#14 0x000000000042ae89 in testing::internal::HandleSehExceptionsInMethodIfSuppor---Type <return> to continue, or q <return> to quit---ted<testing::Test, void> (object=object@entry=0x8d3fd8,
    method=<optimized out>, location=location@entry=0x5b2201 "the test body")
    at common/gtest-all.cc:3562
#15 0x000000000042d60a in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=object@entry=0x8d3fd8,
    method=&virtual testing::Test::TestBody(),
    location=location@entry=0x5b2201 "the test body")
    at common/gtest-all.cc:3598
#16 0x0000000000420bcc in testing::Test::Run (this=this@entry=0x8d3fd8)
    at common/gtest-all.cc:3635
#17 0x0000000000420ca9 in testing::TestInfo::Run (this=0x8733f0)
    at common/gtest-all.cc:3812
#18 0x0000000000420d9b in testing::TestCase::Run (this=0x8c7a02)
    at common/gtest-all.cc:3930
#19 0x0000000000421009 in testing::internal::UnitTestImpl::RunAllTests (
    this=0x852740) at common/gtest-all.cc:5802
#20 0x000000000042b586 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=object@entry=0x852740,
    method=<optimized out>,
    location=location@entry=0x5b35b8 "auxiliary test code (environments or event listeners)") at common/gtest-all.cc:3562
#21 0x000000000042da34 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x852740, method=
---Type <return> to continue, or q <return> to quit---    (bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x420e00 <testing::internal::UnitTestImpl::RunAllTests()>,
    location=location@entry=0x5b35b8 "auxiliary test code (environments or event listeners)") at common/gtest-all.cc:3598
#22 0x0000000000421244 in testing::UnitTest::Run (
    this=0x84b940 <testing::UnitTest::GetInstance()::instance>)
    at common/gtest-all.cc:5416
#23 0x000000000042fa21 in RUN_ALL_TESTS () at common/gtest.h:20058
#24 main (argc=1, argv=0x7fffffff7048) at common/main.cc:79
(gdb) (gdb) frame 6
#6  0x00002aaaab21a0ae in uct_mm_mkey_pack (pd=<optimized out>,
    memh=<optimized out>, rkey_buffer=0x8c79ea) at sm/mm/mm_pd.c:140
140	        strcpy(rkey->path, seg->path);
(gdb)
```

**Comment by shamisp** [comment]
@yosefe general thought - making the choice on per memory block access is a right choice. Nevertheless, for most cases for the same EP the choice is going to be virtually the same. So I think it might be helpful to cache it on EP.

**Comment by yosefe** [comment]
@shamisp not sure additional "if" and lookup would be much faster. But will check.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1883/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1268/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1885/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1270/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1273/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@shamisp I did some performance testing for this PR:
- compared to master, latency is not changed, and message rate is down by 20%;
- when compiling the code with sse4.2 support, latency is up by 4nsec, and message rate is the same
- in both cases, with sse4.2, latency is by 13-17 nsec lower than without it.
The reason is that sse4.2 adds popcount instruction, which is used in rkey lookup. Without it, gcc installs a slower software implementation (sse4.2 is supported starting with Nehalem).

Reminder - we disabled sse by default in order to be compatible across different CPU models.

**Comment by miked-mellanox** [comment]
@yosefe - IMO we should enable sse on ALL, now that we have human error if sse is unsupported at runtime.

**Comment by yosefe** [comment]
Yep, i think so too, because it gives best performance
IMHO we can enable at least SSE4.2 and disable AVX (which is available on Haswell and newer)
We can use a different configuration when building for distros.
@shamisp ?

**Comment by miked-mellanox** [comment]
btw, we can replace human error to fallback. it will try AVX/SSE and fallback to general case if unsupported. gcc/icc pragmas do support it and we have allinfra in place to add it.

**Comment by yosefe** [comment]
@shamisp can we merge this?

**Comment by shamisp** [comment]
👍

**Comment by shamisp** [comment]
We should continue the discussion about SSE default behavior.

File: src/ucp/wireup/wireup.c [hunk 4]
```diff
@@ -204,33 +252,49 @@ ucs_status_t ucp_select_transport(ucp_ep_h ep,
         for (ae = address_list; ae < address_list + address_count; ++ae) {
```
Comment (alinask) [inline src/ucp/wireup/wireup.c]: if a higher score already exists and found=1, then no need to check if remote peer is reachable for this transport's device. it won't be chosen anyway...

Comment (yosefe) [inline src/ucp/wireup/wireup.c]: will be fixes in next PR

**Pull request merged**